flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1389,6 +1388,149 @@ private void copyStateDataHandleData(
 		return Tuple2.of(stateInfo.f0, newMetaInfo);
 	}
 
+	private <N, S extends State, SV> RegisteredKeyValueStateBackendMetaInfo<N, SV> migrateStateIfNecessary(
+			StateDescriptor<S, SV> stateDesc,
+			TypeSerializer<N> namespaceSerializer,
+			Tuple2<ColumnFamilyHandle, RegisteredStateMetaInfoBase> stateInfo,
+			@Nullable StateSnapshotTransformer<SV> snapshotTransformer) throws Exception {
+
+		StateMetaInfoSnapshot restoredMetaInfoSnapshot = restoredKvStateMetaInfos.get(stateDesc.getName());
+
+		Preconditions.checkState(
+			restoredMetaInfoSnapshot != null,
+			"Requested to check compatibility of a restored RegisteredKeyedBackendStateMetaInfo," +
+				" but its corresponding restored snapshot cannot be found.");
+
+		Preconditions.checkState(restoredMetaInfoSnapshot.getBackendStateType()
+				== StateMetaInfoSnapshot.BackendStateType.KEY_VALUE,
+			"Incompatible state types. " +
+				"Was [" + restoredMetaInfoSnapshot.getBackendStateType() + "], " +
+				"registered as [" + StateMetaInfoSnapshot.BackendStateType.KEY_VALUE + "].");
+
+		Preconditions.checkState(
+			Objects.equals(stateDesc.getName(), restoredMetaInfoSnapshot.getName()),
+			"Incompatible state names. " +
+				"Was [" + restoredMetaInfoSnapshot.getName() + "], " +
+				"registered with [" + stateDesc.getName() + "].");
+
+		final StateDescriptor.Type restoredType =
+			StateDescriptor.Type.valueOf(
+				restoredMetaInfoSnapshot.getOption(
+					StateMetaInfoSnapshot.CommonOptionsKeys.KEYED_STATE_TYPE));
+
+		if (!Objects.equals(stateDesc.getType(), StateDescriptor.Type.UNKNOWN)
+			&& !Objects.equals(restoredType, StateDescriptor.Type.UNKNOWN)) {
+
+			Preconditions.checkState(
+				stateDesc.getType() == restoredType,
+				"Incompatible key/value state types. " +
+					"Was [" + restoredType + "], " +
+					"registered with [" + stateDesc.getType() + "].");
+		}
+
+		TypeSerializer<SV> stateSerializer = stateDesc.getSerializer();
+
+		RegisteredKeyValueStateBackendMetaInfo<N, SV> newMetaInfo = new RegisteredKeyValueStateBackendMetaInfo<>(
+			stateDesc.getType(),
+			stateDesc.getName(),
+			namespaceSerializer,
+			stateSerializer,
+			snapshotTransformer);
+
+		CompatibilityResult<N> namespaceCompatibility = CompatibilityUtil.resolveCompatibilityResult(
```

Comment:
Yes, this indeed is not nice. We actually have a newer replacement class called `TypeSerializerSchemaCompatibility`, but at the moment the use of `CompatibilityResult` can't be completely avoided yet.

I have been discussing with @dawidwys that we maybe should just remove `CompatibilityResult` completely in favor of the newer replacement, but this might come up as a separate commit / PR.

Reply:
Yes, this indeed is not nice. We actually have a newer replacement class called `TypeSerializerSchemaCompatibility`, but at the moment the use of `CompatibilityResult` can't be completely avoided yet.

I have been discussing with @dawidwys that we maybe should just remove `CompatibilityResult` completely in favor of the newer replacement, but this might come up as a separate commit / PR.